Workaround ETCD Snapshots Part Files Issue

mattmattox opened the pull request:
To workaround issue [gh-30662](https://github.com/rancher/rancher/issues/30662) please one of the following deployments.

dnoland1 reviewed (changes requested):
I tested the `alt-s3-sync.yaml` solution on a 6 node cluster. It had 3 etcd nodes and 3 control plane/worker nodes. The CronJob object was created, a pod was scheduled, but the pod failed to run successfully. Received the following in the pod logs:
```
time="2021-01-11T21:50:04Z" level=info msg="invoking set s3 service client" s3-accessKey=AKIAQNIJOS7WCY6EV46V s3-bucketName=dnoland-s3-backup s3-endpoint=s3-us-west-2.amazonaws.com s3-endpoint-ca= s3-folder=etcd-fix-testing s3-region=us-west-2

time="2021-01-11T21:50:04Z" level=info msg="Initializing Rolling Backups" creation=5m0s retention=24h0m0s

time="2021-01-11T21:55:06Z" level=warning msg="Checking member health failed from etcd member" data="127.0.0.1:2379 is unhealthy: failed to connect: dial tcp 127.0.0.1:2379: connect: connection refused\nError: unhealthy cluster\n" error="exit status 1"
```
This pod was scheduled on one of the control plane/worker nodes and not on the etcd node. Perhaps the YAML needs affinity rules to ensure it runs on etcd, similar to the `delete-part-files.yaml`. Can you have a look?

mattmattox commented:
@dnoland1 I added the missing nodeAffinity rules to force the pod to only run on an etcd node using the label node-role.kubernetes.io/etcd=true
I also added the flag --once to fix the backup looping issue.

dnoland1 commented:
I redeployed with latest `alt-s3-sync.yaml` but it's still failing. My nodes:
```
$ kubectl get nodes
NAME           STATUS   ROLES                 AGE   VERSION
ip-10-0-7-10   Ready    etcd                  3d    v1.16.15
ip-10-0-7-11   Ready    etcd                  3d    v1.16.15
ip-10-0-7-12   Ready    etcd                  3d    v1.16.15
ip-10-0-7-13   Ready    controlplane,worker   3d    v1.16.15
ip-10-0-7-14   Ready    controlplane,worker   3d    v1.16.15
ip-10-0-7-15   Ready    controlplane,worker   3d    v1.16.15
```
Pod getting scheduled on ip-10-0-7-15 and failing because it's not an etcd node:
```
$ kubectl get pods -n kube-system -o wide | grep alt-etcd
alt-etcd-s3-sync-1610654400-nvpqc         1/1     Running     0          15s   10.0.7.15   ip-10-0-7-15   <none>           <none>
alt-etcd-s3-sync-1610654400-q6sht         0/1     Error       0          88s   10.0.7.15   ip-10-0-7-15   <none>           <none>

```

dnoland1 commented:
Also tested the `delete-part-files.yaml` and was unable to deploy it:
```
$ kubectl apply -f delete-part-files.yaml
error: error validating "delete-part-files.yaml": error validating data: ValidationError(DaemonSet.spec.template.spec.affinity.nodeAffinity.preferredDuringSchedulingIgnoredDuringExecution[0]): missing required field "weight" in io.k8s.api.core.v1.PreferredSchedulingTerm; if you choose to ignore these errors, turn validation off with --validate=false
```

dnoland1 commented:
The cleanup deamonset is running on the wrong nodes:
```
$ kubectl get pods -n kube-system -o wide -l app=cleanup-etcd
NAME                 READY   STATUS    RESTARTS   AGE     IP          NODE           NOMINATED NODE   READINESS GATES
cleanup-etcd-65wf2   1/1     Running   0          7m38s   10.42.4.6   ip-10-0-7-13   <none>           <none>
cleanup-etcd-jqltq   1/1     Running   0          7m38s   10.42.5.4   ip-10-0-7-15   <none>           <none>
cleanup-etcd-kbrps   1/1     Running   0          7m38s   10.42.0.7   ip-10-0-7-14   <none>           <none>
$ kubectl get nodes
NAME           STATUS   ROLES                 AGE    VERSION
ip-10-0-7-10   Ready    etcd                  4d4h   v1.16.15
ip-10-0-7-11   Ready    etcd                  4d4h   v1.16.15
ip-10-0-7-12   Ready    etcd                  4d4h   v1.16.15
ip-10-0-7-13   Ready    controlplane,worker   4d3h   v1.16.15
ip-10-0-7-14   Ready    controlplane,worker   4d4h   v1.16.15
ip-10-0-7-15   Ready    controlplane,worker   4d3h   v1.16.15
```
Should be running on 3 etcd nodes, not 3 controlplane,worker nodes.

dnoland1 reviewed (changes requested):
Tested delete-part-files.yaml. Pods were deployed on etcd nodes and cleanup removed the test .part file I created:
```
Before cleanup
total 660K
-rw-------    1 root     root      657.3K Jan 11 19:16 c-m54tj-rl-v6ldx_2021-01-11T19:15:59Z.zip
-rw-r--r--    1 root     root           0 Jan 14 19:41 dummy-test.part
Post cleanup
total 660K
-rw-------    1 root     root      657.3K Jan 11 19:16 c-m54tj-rl-v6ldx_2021-01-11T19:15:59Z.zip
Sleeping...
```

Tested alt-etcd-s3-sync:
```
time="2021-01-22T16:00:10Z" level=info msg="Initializing Onetime Backup" name=etcd-rolling-snapshots
time="2021-01-22T16:00:11Z" level=info msg="Created local backup" name=etcd-rolling-snapshots runtime=135.793821ms
time="2021-01-22T16:00:11Z" level=info msg="invoking uploading backup file [etcd-fix-testing/etcd-rolling-snapshots.zip] to s3"
time="2021-01-22T16:00:11Z" level=info msg="Successfully uploaded [etcd-fix-testing/etcd-rolling-snapshots.zip] of size [802453]"
```
And see file in S3:
![image](https://user-images.githubusercontent.com/468462/105514772-c1b36800-5c88-11eb-8466-95344457b336.png)

However, the next snapshot overwrites the earlier one:
![image](https://user-images.githubusercontent.com/468462/105521228-7735e980-5c90-11eb-9976-90d2c644f00b.png)


Should use a unique name when uploading the snapshots otherwise only the most recent snapshot is stored in S3.

mattmattox commented:
@dnoland1 Ok. In order to get the names in etcd working correctly IE same as Rancher normally works. I had to switch from using a k8s cronjob to a standard deployment. It nows just uses the built-in scheduler inside rke-tools.

dnoland1 reviewed (approved):
Looks good, seeing multiple backups on S3 now:
![image](https://user-images.githubusercontent.com/468462/105774761-9051cb00-5f1a-11eb-8549-0161507cff6f.png)
